mmdet/models/trainer.py

```diff
@@ -0,0 +1,68 @@
+# Copyright (c) Open-MMLab. All rights reserved.
+import torch
+from mmcv.parallel.data_container import DataContainer
+from mmcv.parallel import MMDataParallel
+
+
+def scatter_cpu(inputs, target_gpus, dim=0):
+    """Scatter inputs to cpu.
+    :type:`~mmcv.parallel.DataContainer`.
+    """
+
+    def scatter_map(obj):
+        if isinstance(obj, torch.Tensor):
+            return [obj]
+        if isinstance(obj, DataContainer):
+            return obj.data
+        if isinstance(obj, tuple) and len(obj) > 0:
+            return list(zip(*map(scatter_map, obj)))
+        if isinstance(obj, list) and len(obj) > 0:
+            out = list(map(list, zip(*map(scatter_map, obj))))
+            return out
+        if isinstance(obj, dict) and len(obj) > 0:
+            out = list(map(type(obj), zip(*map(scatter_map, obj.items()))))
+            return out
+        return [obj for targets in target_gpus]
+
+    # After scatter_map is called, a scatter_map cell will exist. This cell
+    # has a reference to the actual function scatter_map, which has references
+    # to a closure that has a reference to the scatter_map cell (because the
+    # fn is recursive). To avoid this reference cycle, we set the function to
+    # None, clearing the cell
+    try:
+        return scatter_map(inputs)
+    finally:
+        scatter_map = None
+
+
+def scatter_kwargs(inputs, kwargs, target_gpus, dim=0):
+    """Scatter with support for kwargs dictionary"""
+    inputs = scatter_cpu(inputs, target_gpus, dim) if inputs else []
+    kwargs = scatter_cpu(kwargs, target_gpus, dim) if kwargs else []
+    if len(inputs) < len(kwargs):
+        inputs.extend([() for _ in range(len(kwargs) - len(inputs))])
+    elif len(kwargs) < len(inputs):
+        kwargs.extend([{} for _ in range(len(inputs) - len(kwargs))])
+    inputs = tuple(inputs)
+    kwargs = tuple(kwargs)
+    return inputs, kwargs
+
+
+class MMDataCPU(MMDataParallel):
+
+    def scatter(self, inputs, kwargs, device_ids):
```

Comment:
What is `device_ids`? How many devices are supported for a single training?

Reply:
> How many devices are supported for a single training run?

I hope that only one.
Removed `device_ids` and `target_gpus`.